smdebug/core/hook.py

```diff
@@ -456,6 +461,10 @@ def _write_state(self):
             self.state_store.update_state(current_state)
 
     def set_mode(self, mode):
+        # flush out any writes before switching modes
+        if self.writer is not None:
```

Comment:
Close writers regardless of self.writer, as in TF Distributed the writers could be in self.writer_map. Also there could be tbwriters we want to close.

Reply:
calling close_writers here without the check (the very first time) leads to the hang-like behavior that we observed before when close_writer is called the first time in forward_pre_hook()

Reply:
we dont need to close writers anymore because mode is saved in scalar object